L58-python-async-api.md

```diff
@@ -0,0 +1,453 @@
+
+Changes in `grpc.aio.channel_ready_future`:
+* Accepts a `grpc.aio.Channel`;
+* Returns a coroutine object.
+
+### Shared APIs
+
+APIs in the following categories remain in top level:
+* Credentials related classes including for channel, call, and server;
+* Channel connectivity Enum class;
+* Status code Enum class;
+* Compression method Enum class;
+* `grpc.RpcError` exception class;
+* `grpc.RpcContext` class;
+* `grpc.ServiceRpcHandler` class;
+
+### Re-Use ProtoBuf Generated Code For Services
+
+Here is an example of the core part of the gRPC ProtoBuf plugin generated code:
+
+```Python
+# Client-side stub
+class GreeterStub(object):
+  def __init__(self, channel):
+    self.SayHello = channel.unary_unary(...)
+
+# Server-side servicer
+def add_GreeterServicer_to_server(servicer, server):
+  ...
+```
+
+As you can see, the `channel` and `server` object are passed in by the
+application. Thanks to the dynamic nature of Python, the generated code is
+agnostic to which set of API the application uses, as long as we respect the
+method name in our API contract.
+
+### API Reference Doc Generation
+
+The async API related classes and functions, including shared APIs, should be
+generated on the same page.
+
+## Controversial Details
+
+### Generator Or Asynchronous Generator?
+
+For streaming calls, the requests on the client-side are supplied by generator
+for now. If a user wants to provide a pre-defined list of request messages, they
+can use build-in `iter()` function. But there isn't an equivalent function for
+async generator. Should we wrap it inside our library to increase usability?
+
+```Python
+### Current usage of Python generator
+stub.SayHelloStreaming(iter([
+    HelloRequest(name='Golden'),
+    HelloRequest(name='Retriever'),
+    HelloRequest(name='Pan'),
+    HelloRequest(name='Cake'),
+]))
+
+### The new usage is much verbose for same scenario
+class AsyncIter:
+    def __init__(self, items):    
+        self.items = items    
+
+    async def __aiter__(self):    
+        for item in self.items:    
+            yield item
+
+stub.SayHelloStreaming(AsyncIter([
+    HelloRequest(name='Golden'),
+    HelloRequest(name='Retriever'),
+    HelloRequest(name='Pan'),
+    HelloRequest(name='Cake'),
+]))
+```
+
+### Special Async Functions Name
+
+Fire-and-forget is a valid use case in async programming. For the non-critical
+tasks, the program may schedule the execution without checking the result. As
+mentioned in "Python Coroutines in `asyncio`" section, the coroutine won't be
+scheduled unless we explicitly do so. A dedicated prefix or suffix of the
+function should help to remind developers to await the coroutine.
+
+```Python
+### Forget to await RPC
+while 1:
+    stub.ReportLoad(ReportRequest(
+        timestamp=...,
+        metrics=[...],
+    ))  # <- RPC not sent
+    await asyncio.sleep(3)
+
+### Await on whatever function starts with "Async"
+while 1:
+    await stub.AsyncReportLoad(ReportRequest(
```

Comment:
I'm also in favor of not making modification. I'm mentioning this is because people do use this naming pattern when they adopt `asyncio` (like [Instagram](https://www.slideshare.net/jimmy_lai/the-journey-of-asyncio-adoption-in-instagram)). 

Now, I'm thinking maybe we can name the service stub `async_stub` in our example code.